### PR TITLE
breakdown balances per protocol

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Part of #4455

### CODE REVIEW

File: docs/api.rst [hunk 1]
```diff
@@ -2985,11 +2985,20 @@ Querying onchain balances
...
                    "eth": { "0x78b0AD50E768D2376C6BA7de33F426ecE4e03e0B": {
                        "assets": {
-                           "ETH": {"amount": "10", "usd_value": "1650.53"},
-                           "eip155:1/erc20:0x6B175474E89094C44Da98b954EedeAC495271d0F": {"amount": "15", "usd_value": "15.21"}
+                           "ETH": {
+                               "ACCOUNT": {"amount": "10", "usd_value": "1650.53"}
+                           },
+                           "eip155:1/erc20:0x6B175474E89094C44Da98b954EedeAC495271d0F": {
+                               "ACCOUNT": {"amount": "15", "usd_value": "15.21"}
+                           },
```
Comment (yabirgb) [inline docs/api.rst]: maybe `wallet` instead of `account` is a better name?

File: rotkehlchen/accounting/structures/balance.py [hunk 2]
```diff
@@ -141,29 +141,50 @@ def serialize_for_db(self) -> tuple[str, str, str]:
...
     def to_dict(self) -> dict[str, dict]:
         return {
-            'assets': {k: v.to_dict() for k, v in self.assets.items()},
-            'liabilities': {k: v.to_dict() for k, v in self.liabilities.items()},
+            'assets': {
+                k: {protocol: balance.to_dict() for protocol, balance in v.items()}
+                for k, v in self.assets.items()
+            },
+            'liabilities': {
+                k: {protocol: balance.to_dict() for protocol, balance in v.items()}
```
Comment (yabirgb) [inline rotkehlchen/accounting/structures/balance.py]: it is a nitpick but reading protocol here is confusing since it cal also be the in the account itself. Maybe using something like `location_label` is better but it can be confused with our location_label from history events. What about just label?

File: rotkehlchen/chain/aggregator.py [hunk 3]
```diff
@@ -666,9 +669,8 @@ def query_ksm_balances(
                 amount=amount,
                 usd_value=amount * ksm_usd_price,
             )
-            self.balances.ksm[account] = BalanceSheet(
-                assets=defaultdict(Balance, {A_KSM: balance}),
-            )
+            self.balances.ksm[account] = BalanceSheet()
+            self.balances.ksm[account].assets[A_KSM]['ACCOUNT'] = balance
```
Comment (yabirgb) [inline rotkehlchen/chain/aggregator.py]: 'ACCOUNT' should be a constant since you use a lot of times

File: rotkehlchen/chain/aggregator.py [hunk 4]
```diff
@@ -911,16 +912,16 @@ def query_eth2_balances(
...
+            self.balances.eth2[pubkey].assets[A_ETH2]['ACCOUNT'] = balance
 
+    @staticmethod
     def _update_balances_after_token_query(
-            self,
             dsr_proxy_append: bool,
             balance_result: dict[ChecksumEvmAddress, dict[EvmToken, FVal]],
             token_usd_price: dict[EvmToken, Price],
             balances: defaultdict[ChecksumEvmAddress, BalanceSheet],
+            protocol_or_account: Literal['ACCOUNT', 'makerdao vault'] = 'ACCOUNT',
```
Comment (yabirgb) [inline rotkehlchen/chain/aggregator.py]: if we find a better name we can use it instead of protocol_or_account

File: rotkehlchen/chain/balances.py [hunk 5]
```diff
@@ -125,9 +125,9 @@ def recalculate_totals(self) -> BalanceSheet:
                 new_totals += chain_balances
 
         for btc_balance in self.btc.values():
-            new_totals.assets[A_BTC] += btc_balance
+            new_totals.assets[A_BTC]['ACCOUNT'] += btc_balance
         for bch_balance in self.bch.values():
-            new_totals.assets[A_BCH] += bch_balance
+            new_totals.assets[A_BCH]['ACCOUNT'] += bch_balance
```
Comment (yabirgb) [inline rotkehlchen/chain/balances.py]: use a constant

File: rotkehlchen/chain/ethereum/interfaces/balances.py [hunk 6]
```diff
@@ -242,13 +242,15 @@ def query_balances(self) -> BalancesSheetType:
...
-            balances[address].assets = self._query_gauges_balances(
+            gauge_balances = self._query_gauges_balances(
                 user_address=address,
                 gauges_to_token=gauges_to_token,
                 call_order=call_order,
                 chunk_size=chunk_size,
                 balances_contract=self._get_staking_contract_balances,
             )
+            for asset, balance in gauge_balances.items():
+                balances[address].assets[asset][self.counterparty] += balance
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/interfaces/balances.py]: `_query_gauges_balances` is used only here. Modify it to set the counterparty there instead of looping once again just to adjust for it

File: rotkehlchen/rotkehlchen.py [hunk 7]
```diff
@@ -1035,9 +1035,28 @@ def query_balances(
                 blockchain=None,
                 ignore_cache=ignore_cache,
             )  # copies below since if cache is used we end up modifying the balance sheet object
-            if len(blockchain_result.totals.assets) != 0:
-                balances[str(Location.BLOCKCHAIN)] = blockchain_result.totals.assets.copy()
-            liabilities = blockchain_result.totals.liabilities.copy()
+
+            # Convert nested assets structure to flat dict for backward compatibility
```
Comment (yabirgb) [inline rotkehlchen/rotkehlchen.py]: why do you need backwards compatibility? Where is this one used?

File: rotkehlchen/rotkehlchen.py [hunk 8]
```diff
@@ -1070,7 +1089,6 @@ def query_balances(
                 if len(loopring_balances) != 0:
                     balances[str(Location.LOOPRING)] = loopring_balances
 
-        # retrieve nft balances if module is activated
```
Comment (yabirgb) [inline rotkehlchen/rotkehlchen.py]: why remove this comment?

File: rotkehlchen/tests/api/blockchain/test_avalanche.py [hunk 9]
```diff
@@ -78,13 +78,13 @@ def test_add_avax_blockchain_account(rotkehlchen_api_server: 'APIServer') -> Non
...
     assert 'liabilities' in account_balances
-    asset_avax = account_balances['assets'][A_AVAX.identifier]
+    asset_avax = account_balances['assets'][A_AVAX.identifier]['ACCOUNT']
     assert FVal(asset_avax['amount']) >= ZERO
     assert FVal(asset_avax['usd_value']) >= ZERO
 
     # Check totals
     assert 'liabilities' in result['totals']
-    total_avax = result['totals']['assets'][A_AVAX.identifier]
+    total_avax = result['totals']['assets'][A_AVAX.identifier]['ACCOUNT']
```
Comment (yabirgb) [inline rotkehlchen/tests/api/blockchain/test_avalanche.py]: use the constant

File: rotkehlchen/tests/api/blockchain/test_optimism.py [hunk 10]
```diff
@@ -53,13 +53,13 @@ def test_add_optimism_blockchain_account(rotkehlchen_api_server: 'APIServer') ->
     # Check per account
     account_balances = result['per_account'][optimism_chain_key][TEST_ADDY]
     assert 'liabilities' in account_balances
-    asset_eth = account_balances['assets'][A_ETH.identifier]
+    asset_eth = account_balances['assets'][A_ETH.identifier]['ACCOUNT']
```
Comment (yabirgb) [inline rotkehlchen/tests/api/blockchain/test_optimism.py]: in all this files please use a constant for 'ACCOUNT'

File: rotkehlchen/chain/aggregator.py (hunk 4 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/aggregator.py]: `balance_label`. i guess that makes more sense.

File: rotkehlchen/rotkehlchen.py (hunk 8 above)
Comment (prettyirrelevant) [reply rotkehlchen/rotkehlchen.py]: mistake. added it back.

File: rotkehlchen/rotkehlchen.py (hunk 7 above)
Comment (prettyirrelevant) [reply rotkehlchen/rotkehlchen.py]: sorry, that shouldn't be here. this is used to take balance snapshots.

File: rotkehlchen/tests/unit/test_blockchain_balances.py [hunk 11]
```diff
@@ -196,14 +195,14 @@ def test_protocol_balances(blockchain: 'ChainsAggregator') -> None:
     blockchain._add_eth_protocol_balances(eth_balances=blockchain.balances.eth)
     # the proxy balances are added to the owner's
     assert blockchain.balances.eth[ETH_ADDRESS1].assets == {
-        A_LQTY: Balance(
+        A_LQTY: {'liquity': Balance(
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_blockchain_balances.py]: use the counterparty

```suggestion
        A_LQTY: {CPT_LIQUITY: Balance(
```

File: rotkehlchen/tests/unit/test_blockchain_balances.py [hunk 12]
```diff
@@ -196,14 +195,14 @@ def test_protocol_balances(blockchain: 'ChainsAggregator') -> None:
...
             amount=FVal('16535.272316119505457412'),
             usd_value=FVal('24802.9084741792581861180'),
-        ),
-        A_LUSD: Balance(
+        )},
+        A_LUSD: {'liquity': Balance(
             amount=FVal('58774.021313242937366493'),
             usd_value=FVal('88161.0319698644060497395'),
-        ),
+        )},
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_blockchain_balances.py]: same, use the counterparty

File: rotkehlchen/tests/unit/test_makerdao.py [hunk 13]
```diff
@@ -161,8 +161,8 @@ def test_get_vault_balance(
         stability_fee=ZERO,
     )
     expected_result = BalanceSheet(
-        assets=defaultdict(Balance, {A_ETH: Balance(FVal('100'), FVal('20000'))}),
-        liabilities=defaultdict(Balance, {A_DAI: Balance(FVal('2000'), FVal('2020'))}),
+        assets=defaultdict(Balance, {A_ETH: {'makerdao vault': Balance(FVal('100'), FVal('20000'))}}),  # noqa: E501
+        liabilities=defaultdict(Balance, {A_DAI: {'makerdao vault': Balance(FVal('2000'), FVal('2020'))}}),  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_makerdao.py]: use the counterparty variables so it is easier to follow where they are used and also so they can be changed

File: rotkehlchen/tests/unit/test_protocol_balances.py [hunk 14]
```diff
@@ -295,7 +295,7 @@ def test_thegraph_balances(
     )
     thegraph_balances = thegraph_balances_inquirer.query_balances()
     user_balance = thegraph_balances[ethereum_accounts[0]]
-    assert user_balance.assets[A_GRT.resolve_to_evm_token()] == Balance(
+    assert user_balance.assets[A_GRT.resolve_to_evm_token()][thegraph_balances_inquirer.counterparty] == Balance(  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_protocol_balances.py]: 2 things:

1. resolve once to evm token
2. if you use CPT_GRAPH or whatever is the variable can you avoid the noqa? Even if not I would use CPT_THEGRAPH

File: rotkehlchen/tests/unit/test_protocol_balances.py [hunk 15]
```diff
@@ -1094,7 +1069,7 @@ def test_umami_balances(
     )
     protocol_balances = protocol_balances_inquirer.query_balances()
     user_balance = protocol_balances[arbitrum_one_accounts[0]]
-    assert user_balance.assets[gm_usdc_vault] == Balance(
+    assert user_balance.assets[gm_usdc_vault][protocol_balances_inquirer.counterparty] == Balance(
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_protocol_balances.py]: same, I would use here directly the variable instead of accessing the attribute of the object

File: rotkehlchen/utils/misc.py [hunk 16]
```diff
@@ -175,6 +175,28 @@ def combine_dicts(
...
+            new_dict[key] = combine_dicts(new_dict[key], inner_dict, op)
+        elif op == operator.sub:  # pylint: disable=comparison-with-callable
+            new_inner: defaultdict[Any, V] = defaultdict(inner_dict.default_factory)
+            for k, v in inner_dict.items():
+                new_inner[k] = -v  # type: ignore
+            new_dict[key] = new_inner
+        else:
+            new_dict[key] = inner_dict.copy()
+
+    return new_dict
```
Comment (yabirgb) [inline rotkehlchen/utils/misc.py]: I'm worried about the performance of this function since you copy the dict. Do you need to copy at all or can you replace in place instead?

**Comment by yabirgb** [comment]
The failing test is not related and isaac is looking at it. It fails sometimes

File: docs/api.rst [hunk 17]
```diff
@@ -2985,11 +2985,20 @@ Querying onchain balances
                    },
                    "eth": { "0x78b0AD50E768D2376C6BA7de33F426ecE4e03e0B": {
                        "assets": {
-                           "ETH": {"amount": "10", "usd_value": "1650.53"},
-                           "eip155:1/erc20:0x6B175474E89094C44Da98b954EedeAC495271d0F": {"amount": "15", "usd_value": "15.21"}
+                           "ETH": {
+                               "wallet": {"amount": "10", "usd_value": "1650.53"}
+                           },
+                           "eip155:1/erc20:0x6B175474E89094C44Da98b954EedeAC495271d0F": {
+                               "wallet": {"amount": "15", "usd_value": "15.21"}
```
Comment (LefterisJP) [inline docs/api.rst]: Can we change that to "address" instead of wallet? As wallet is the software that the user runs.

File: docs/api.rst [hunk 18]
```diff
@@ -3017,7 +3026,7 @@ Querying onchain balances
...
+   :resjson object per_account: The blockchain balances per account per asset. Each element of this object has a blockchain asset as its key. Then each asset has an address for that blockchain as its key and each address an object with the following keys: ``"amount"`` for the amount stored in the asset in the address and ``"usd_value"`` for the equivalent USD value as of the request. For EVM accounts, balances are organized by category: ``"wallet"`` represents tokens held in the wallet and available for use, while protocol-specific labels (like ``"makerdao vault"``, ``"aave"``, ``"morpho"``) show tokens locked in DeFi protocols. ETH accounts may have an optional liabilities key with the same structure. BTC accounts are separated in standalone accounts and in accounts that have been derived from an xpub. The xpub ones are listed in a list under the ``"xpubs"`` key. Each entry has the xpub, the derivation path and the list of addresses and their balances.
```
Comment (LefterisJP) [inline docs/api.rst]:
```suggestion
   :resjson object per_account: The blockchain balances per account per asset. Each element of this object has a blockchain asset as its key. Then each asset has an address for that blockchain as its key and each address an object with the following keys: ``"amount"`` for the amount stored in the asset in the address and ``"usd_value"`` for the equivalent USD value as of the request. For EVM accounts, balances are organized by category: ``"address"`` represents tokens held directly in the address and available for use, while protocol-specific labels (like ``"makerdao vault"``, ``"aave"``, ``"morpho"``) show tokens locked in DeFi protocols. ETH accounts may have an optional liabilities key with the same structure. BTC accounts are separated in standalone accounts and in accounts that have been derived from an xpub. The xpub ones are listed in a list under the ``"xpubs"`` key. Each entry has the xpub, the derivation path and the list of addresses and their balances.
```

File: rotkehlchen/constants/misc.py [hunk 19]
```diff
@@ -37,3 +37,5 @@
 APPDIR_NAME: Final = 'app'
 AIRDROPSDIR_NAME: Final = 'airdrops'
 AIRDROPSPOAPDIR_NAME: Final = 'airdrops_poap'
+
+DEFAULT_BALANCE_LABEL: Final = 'wallet'
```
Comment (LefterisJP) [inline rotkehlchen/constants/misc.py]: As I said let's change that to `'address'`

File: rotkehlchen/tests/unit/test_evm_balances.py [hunk 20]
```diff
@@ -48,4 +50,7 @@ def test_optimism_balances(
         ignore_cache=True,
     )
     account_balance = blockchain_result.per_account.optimism[optimism_accounts[0]].assets[velodrome_v2_weth_op_lp_token]  # noqa: E501
-    assert account_balance.amount == FVal('0.086313645974870917')  # 0.043225873904215354 in wallet + 0.043087772070655563 staked  # noqa: E501
+    assert account_balance == {
+        DEFAULT_BALANCE_LABEL: Balance(amount=FVal('0.043225873904215354'), usd_value=FVal('0.0648388108563230310')),  # noqa: E501
+        'velodrome': Balance(amount=FVal('0.043087772070655563'), usd_value=FVal('0.0646316581059833445')),  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_evm_balances.py]: use the CPT constant here

File: rotkehlchen/utils/misc.py [hunk 21]
```diff
@@ -175,6 +175,26 @@ def combine_dicts(
     return new_dict
 
 
+def combine_nested_dicts_inplace(
```
Comment (LefterisJP) [inline rotkehlchen/utils/misc.py:178]: Can you please write a small unit test for this function. Basic usage and edge cases.

File: rotkehlchen/utils/misc.py [hunk 22]
```diff
@@ -175,6 +175,26 @@ def combine_dicts(
...
+            a[key] = combine_dicts(a[key], inner_dict, op)
+        elif op == operator.sub:  # pylint: disable=comparison-with-callable
+            new_inner: defaultdict[Any, V] = defaultdict(inner_dict.default_factory)
+            for k, v in inner_dict.items():
+                new_inner[k] = -v  # type: ignore
+            a[key] = new_inner
+        else:
+            a[key] = inner_dict.copy()
+
+    return a
```
Comment (LefterisJP) [inline rotkehlchen/utils/misc.py]: Isn't this a bit more efficient implementation by not reusing combine_nested_dicsts? 

```suggestion
      """                                                                                                              
      Combines nested defaultdicts by modifying `a` in place.
                                                          
      For each key in `b`, combines the inner dicts using the given operation.
      Returns the modified `a` for method chaining.                                                                    
      """
      for outer_key, inner_dict_b in b.items():                                                                        
          if outer_key not in a:
              # Create new inner dict with same default factory as b's inner dict
              a[outer_key] = defaultdict(inner_dict_b.default_factory)
                                                           
          inner_dict_a = a[outer_key]            

          # Combine inner dictionaries
          for inner_key, value_b in inner_dict_b.items():
              if inner_key in inner_dict_a:
                  inner_dict_a[inner_key] = op(inner_dict_a[inner_key], value_b)
              else:
                  # For subtraction, negate the value since it's not in a
                  if op is operator.sub:
                      inner_dict_a[inner_key] = op(inner_dict_a.default_factory(), value_b)
                  else:
                      inner_dict_a[inner_key] = value_b

      return a
```

**Review by LefterisJP — APPROVED** [review]
LGTM
